### DISCUSSION TITLE
Django APscheduler prevent more workers running scheduled task

### DISCUSSION THREAD

**Discussion by MarkoKlima** [agronholm/apscheduler#592] [Q&A]
I use APScheduler in Django, on Windows IIS (.NET CLR Version v4.0, AlwaysRunning, Maximum Worker Processes 5) to run my background script. Problem is, taks gets run multiple times. If I run same program on my PC, it only runs once, but when I upload to windows server (which hosts my Django app) it runs more times. I guess it has some connection with the number of workers? Job is scheduled, but each time job task is done, it's like it runs random number of instances. First 1 time, then 2, then 10, then again 2. Even tho I have 'replace_existing=True, coalesce= True, misfire_grace_time = 1, max_instances = 1'

planer_zad.py
```
from apscheduler.schedulers.background import BackgroundScheduler
from blog.views import cron_mail_overdue

def start():

scheduler.add_job(cron_mail_overdue, "cron", hour=7, minute=14, day_of_week='mon-sun', id="task002", replace_existing=True, coalesce= True, misfire_grace_time = 10, max_instances = 1)

scheduler.start()
```
apps.py
```
from django.apps import AppConfig


class BlogConfig(AppConfig):
    name = 'blog'

    def ready(self):
        #print('Starting Scheduler...')
        from .planer import planer_zad
        planer_zad.start()
```
For test I tried 'interval':
```
scheduler.add_job(cron_mail_overdue, "interval", minutes=1, id="task002", replace_existing=True, coalesce= True, misfire_grace_time = 10, max_instances = 1)
```

Tried:
```
    scheduler = BackgroundScheduler({
        'apscheduler.executors.default': {
            'class': 'apscheduler.executors.pool:ThreadPoolExecutor',
            'max_workers': '1'
        },
        'apscheduler.executors.processpool': {
            'type': 'processpool',
            'max_workers': '1'
        },
        'apscheduler.job_defaults.coalesce': 'True',
        'apscheduler.job_defaults.max_instances': '1',
        'apscheduler.timezone': 'UTC',
    })


scheduler.add_job(cron_mail_overdue, "cron", hour=9, minute=3, second=00, day_of_week='mon-sun', id="task002", replace_existing=True, coalesce= True, misfire_grace_time = 10, max_instances = 1)

scheduler.start()
```

Does not work. Sometimes it runs only once, then 12 times. Also, I tried using two jobs, 1 to send email and write BOOL to database (signal that email has been sent), to block sending more then once. Then after 30min, I unblock that BOOL. It didn't work, it also sent 6 emails.

[![Example of received emails][1]][1]

  [1]: https://i.stack.imgur.com/OQldp.png

**Comment by simbakowo** [comment]
Did you figure this out?

**Comment by agronholm** [comment]
Have you read the [FAQ](https://apscheduler.readthedocs.io/en/3.x/faq.html#how-do-i-share-a-single-job-store-among-one-or-more-worker-processes)?

**Comment by Mhammouda** [comment]
Hello @MarkoKlima , i have same issue .
i need a help please

**Reply by agronholm** [reply]
It's better if you make your own case, including whatever you've done. Also, make sure you've read the FAQ to understand the technical restrictions involved here.

**Reply by Mhammouda** [reply]
@agronholm ,its the same case exactly than @MarkoKlima , i look in the FAQ document no line describe  this case

**Reply by Mhammouda** [reply]
@agronholm , if you have any idea about this issue , please share her

**Reply by agronholm** [reply]
If you're running multiple workers, and a scheduler in your app, you're running multiple scheduler instances which may not be what you want. Is this the case? Can you also confirm which release you're running?

**Reply by Mhammouda** [reply]
def test1(serial_number_IAD):
    print("________________________________")
    print(datetime.now())
    print("I M test from ", serial_number_IAD)
    print("________________________________")`

@login_required(login_url='login')
@forCustomer
def rebootloop(request):
    try:
        if request.method == "GET":
            test_ap = AP.objects.filter(user=request.user.username)
            if len(test_ap) == 0:
                messages.warning(request, "ValueError : NO DEVICE ASSIGNED TO YOUR ACCOUNT")
            context = {'test_ap': test_ap}
            return render(request, 'reboot.html', context)
        if request.method == "POST":
            url = request.META.get('HTTP_REFERER')
            serial_number = request.POST.get('monselect')
            test_ap = AP.objects.get(serial_number=serial_number)

            if not test_ap:
                messages.error(request, "ValueError : NO DEVICE ASSIGNED TO YOUR ACCOUNT")
            else:
                last_test = Package.objects.filter(
                    Q(serial_nmber=serial_number) & (Q(status='wait') | Q(status='IN_PROGRESS')))
                if len(last_test) != 0:
                    messages.warning(request, "ValueError : test is already in progress.")
                    return render(request, 'reboot.html')
                else:
                    day = request.POST.get('day')
                    timee = request.POST.get('time')

                    range = (request.POST.get('range'))
                    if range == "" or day == "" or timee == "":
                        messages.error(request, "ValueError :Please complete all fields")
                    else:

                        if test_ap.status == "NOT_CONNECTED":
                            messages.error(request, "ValueError :you cannot run a test on  device Not_Connected")
                        else:

                            range = int(request.POST.get('range'))
                            next_run_time = str(day) + " " + str(timee) + ":00"
                            test_ap.test = "TEST_IN_PROCESSING"
                            test_ap.save()
                            serial_number = test_ap.serial_number
                            current_soft = test_ap.Soft_C0

                            task_id = serial_number + str(next_run_time).replace(".", "").replace("+", "-").replace(":",
                                                                                                                    "-")

                            if range < 1:
                                messages.error(request, "the range of task must be greater than 1")
                            else:
                                try:
                                    logger.info("Starting scheduler...")
                                    scheduler = BackgroundScheduler()
                                    scheduler.add_jobstore(DjangoJobStore(), "default")
                                    time_task = range
                                    scheduler.add_job(

                                        test1, IntervalTrigger(minutes=range),  # Every 10 seconds
                                        id=task_id,  # The `id` assigned to each job MUST be unique
                                        max_instances=100, next_run_time=next_run_time,
                                        replace_existing=False,
                                        args=[serial_number]
                                    )
                                    logger.info("Added job 'my_job'.")
                                    scheduler.start()
                                    messages.success(request, "TASK LANCE")
                                except KeyboardInterrupt:
                                    logger.info("Stopping scheduler...")
                                    scheduler.shutdown()
                                    logger.info("Scheduler shut down successfully!")

                                job = scheduler.get_job(job_id=task_id)

                                filename = "log" + str(serial_number) + str(next_run_time).replace(".", "").replace("+",
                                                                                                                    "-").replace(
                                    ":", "-")

                                b = LOG(serial_nmber=serial_number, taskid=task_id, filename=filename,
                                        user=request.user.username, name_job=job, name=job.name, triger=job.trigger,
                                        time_job=job.next_run_time, nbr_RES=0, nbr_reboot=0, current_soft=current_soft)
                                b.save()

            context = {}
            return redirect(url)

    except:
        messages.error(request, "Pas de Serial pour ce compte")
        return render(request, 'reboot.html')
